### ISSUE TITLE
Correctness of "counter average weight reduction"

### ISSUE THREAD

**Issue by rgdyy** [texttron/tevatron#143]
Hi devs, first, thanks for this clean & useful code base!

I'm not following this line
https://github.com/texttron/tevatron/blob/f88e0c738dd15cfe8cbc939159ef3ed29597eb5a/src/tevatron/retriever/modeling/encoder.py#L72

since in an earlier line https://github.com/texttron/tevatron/blob/f88e0c738dd15cfe8cbc939159ef3ed29597eb5a/src/tevatron/retriever/modeling/encoder.py#L41
reduction is "mean". I believe it is correct to let DDP average the gradient across multiple GPUs, and there's no need to counteract that? Only when DDP does the average, would it be equivalent between 2 GPUs with 8 examples each and 1 GPU with 16 examples, both reduced by "mean".

Thanks in advance.

**Comment by MXueguang** [comment]
Hi @rgdyy ,  https://github.com/texttron/tevatron/issues/30 is relevant.

**Comment by rgdyy** [comment]
Hi @MXueguang , sorry that I forgot about this for a long while. As far as I understood, mean of mean is still correctly, mean, so Line 72 `loss * world_size` should be deleted. What do you think?

**Comment by rgdyy** [comment]
It's been a long time again. I read #30 and also [another thread](https://discuss.pytorch.org/t/torch-distributed-all-gather-to-compute-noise-contrastive-loss-with-pytorch/162015/5). After skimming [the repo](https://github.com/zsnoob/EfficientDDP-4-Contrastive-Train) pointed to from there, I believe the loss scaling is not sufficient: because the gather is non-diff, cross-GPU negatives would receive no grad from the current query's contrastive loss. However, I do not have the time to read and validate all things in that repo, so I would personally choose the solution discussed [here](https://github.com/Lightning-AI/pytorch-lightning/discussions/14390) with PyTorch Lightning. Please let me know if I'm wrong. Thanks for the code and everything!

**Comment by MXueguang** [comment]
Thanks for bring this back. I'll carefully take a look the implements and there affects in model performance recently and let you know @rgdyy
